src/main/java/org/javarosa/xform/parse/FormInstanceParser.java

```diff
@@ -378,68 +386,105 @@ private void verifyItemsetSrcDstCompatibility (FormInstance instance) {
         for (DataBinding bind : bindings) {
-            final TreeReference ref = FormInstance.unpackReference(bind.getReference());
-            final List<TreeReference> nodes = new EvaluationContext(instance).expandReference(ref, true);
-
-            if (nodes.size() > 0) {
-                attachBindGeneral(bind);
-            }
-            for (TreeReference nref : nodes) {
-                attachBind(instance.resolveReference(nref), bind);
+            TreeReference ref = FormInstance.unpackReference(bind.getReference());
+            EvaluationContext ec = new EvaluationContext(instance);
+            List<TreeReference> nodeRefs = ec.expandReference(ref, true);
+
+            // Add triggerable targets if needed
+            if (bind.relevancyCondition != null) {
+                bind.relevancyCondition.addTarget(ref);
+                // Since relevancy can affect not only to individual fields, but also to
+                // groups, we need to register all descendant refs as targets for relevancy
+                // conditions to allow for chained reactions in triggerables registered in
+                // any of those descendants
+                for (TreeReference r : getDescendantRefs(instance, ec, ref))
+                    bind.relevancyCondition.addTarget(r);
+            }
+            if (bind.requiredCondition != null)
+                bind.requiredCondition.addTarget(ref);
+            if (bind.readonlyCondition != null)
+                bind.readonlyCondition.addTarget(ref);
+            if (bind.calculate != null)
+                bind.calculate.addTarget(ref);
+
+            // Initialize present nodes under the provided ref
+            for (TreeReference nodeRef : nodeRefs) {
+                TreeElement node = instance.resolveReference(nodeRef);
+                node.setDataType(bind.getDataType());
+
+                if (bind.relevancyCondition == null)
+                    node.setRelevant(bind.relevantAbsolute);
+                if (bind.requiredCondition == null)
+                    node.setRequired(bind.requiredAbsolute);
+                if (bind.readonlyCondition == null)
+                    node.setEnabled(!bind.readonlyAbsolute);
+                if (bind.constraint != null)
+                    node.setConstraint(new Constraint(bind.constraint, bind.constraintMessage));
+
+                node.setPreloadHandler(bind.getPreload());
+                node.setPreloadParams(bind.getPreloadParams());
+                node.setBindAttributes(bind.getAdditionalAttributes());
             }
         }
 
         applyControlProperties(instance);
     }
 
-    private static void attachBindGeneral (DataBinding bind) {
-        TreeReference ref = FormInstance.unpackReference(bind.getReference());
+    private Set<TreeReference> getDescendantRefs(FormInstance mainInstance, EvaluationContext evalContext, TreeReference original) {
+        // original has already been added to the 'toAdd' list.
```

Comment:
This comment doesn't make sense anymore because there isn't a `toAdd` list. Does the `original` now get added because it is in its own descendant list?

Reply:
`bind.relevancyCondition.addTarget(ref)` right before the call adds the original reference.